components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java

```diff
@@ -104,6 +95,14 @@ public Origin getOrigin() {
     }
 
     private void newConnection() {
+        int borrowed = borrowedCount.get();
+        int inEstablishment = connectionsInEstablishment.getAndIncrement();
```

Comment:
Are we incrementing this value here just to reduce the number of calls to the AtomicInteger? It looks a bit odd that we increment it to decrement it straightaway in some cases, since we don't do the same for the borrowedCount.   I would consider alternative implementations.

Reply:
No. This is to avoid a race condition in relation to `connectionsInEstablishment`. So if we'd do this:

    if (borrowed + connectionsInEstablishment.get() >= max) {
        return;
    } 
    connectionsInEstablishment.increment();
    

Then it is possible (I would say even likely) that several other threads will simultaneously call `get` and pass the test, when they should have failed.

But a call to `getAndIncrement` ensures that each thread sees an updated value and only one (or some) of them will pass the test while others will fail.

However this is not necessary for `borrowCount` because it is never modified in this method. It is modified (atomically) elsewhere whenever a connection is handed out to a borrower, or a connection is returned back to the pool.